File: mediapipe/tasks/cc/text/text_classifier/text_classifier_test.cc
/* Copyright 2022 The MediaPipe Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "mediapipe/tasks/cc/text/text_classifier/text_classifier.h"

#include <memory>
#include <sstream>
#include <string>
#include <utility>

#include "absl/flags/flag.h"
#include "absl/status/status.h"
#include "absl/strings/cord.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"
#include "mediapipe/framework/deps/file_path.h"
#include "mediapipe/framework/port/gmock.h"
#include "mediapipe/framework/port/gtest.h"
#include "mediapipe/framework/port/status_matchers.h"
#include "mediapipe/tasks/cc/common.h"
#include "mediapipe/tasks/cc/components/containers/category.h"
#include "mediapipe/tasks/cc/components/containers/classification_result.h"
#include "mediapipe/tasks/cc/text/text_classifier/text_classifier_test_utils.h"
#include "tensorflow/lite/test_util.h"

namespace mediapipe::tasks::text::text_classifier {
namespace {

using ::mediapipe::file::JoinPath;
using ::mediapipe::tasks::kMediaPipeTasksPayload;
using ::mediapipe::tasks::components::containers::Category;
using ::mediapipe::tasks::components::containers::Classifications;
using ::testing::HasSubstr;
using ::testing::Optional;

constexpr int kMaxSeqLen = 128;
const float kPrecision = 1e-5;
constexpr char kTestDataDirectory[] = "/mediapipe/tasks/testdata/text/";
constexpr char kTestBertModelPath[] = "bert_text_classifier.tflite";
constexpr char kInvalidModelPath[] = "i/do/not/exist.tflite";
constexpr char kTestRegexModelPath[] =
    "test_model_text_classifier_with_regex_tokenizer.tflite";
constexpr char kStringToBoolModelPath[] =
    "test_model_text_classifier_bool_output.tflite";

std::string GetFullPath(absl::string_view file_name) {
  return JoinPath("./", kTestDataDirectory, file_name);
}

// Checks that the two provided `TextClassifierResult` are equal, with a
// tolerancy on floating-point score to account for numerical instabilities.
// TODO: create shared matcher for ClassificationResult.
void ExpectApproximatelyEqual(const TextClassifierResult& actual,
                              const TextClassifierResult& expected) {
  ASSERT_EQ(actual.classifications.size(), expected.classifications.size());
  for (int i = 0; i < actual.classifications.size(); ++i) {
    const Classifications& a = actual.classifications[i];
    const Classifications& b = expected.classifications[i];
    EXPECT_EQ(a.head_index, b.head_index);
    EXPECT_EQ(a.head_name, b.head_name);
    EXPECT_EQ(a.categories.size(), b.categories.size());
    for (int j = 0; j < a.categories.size(); ++j) {
      const Category& x = a.categories[j];
      const Category& y = b.categories[j];
      EXPECT_EQ(x.index, y.index);
      EXPECT_EQ(x.category_name, y.category_name);
      EXPECT_EQ(x.display_name, y.display_name);
    }
  }
}

}  // namespace

class TextClassifierTest : public tflite::testing::Test {};

TEST_F(TextClassifierTest, CreateSucceedsWithBertModel) {
  auto options = std::make_unique<TextClassifierOptions>();
  options->base_options.model_asset_path = GetFullPath(kTestBertModelPath);
  MP_ASSERT_OK(TextClassifier::Create(std::move(options)));
}

TEST_F(TextClassifierTest, CreateFailsWithMissingBaseOptions) {
  auto options = std::make_unique<TextClassifierOptions>();
  StatusOr<std::unique_ptr<TextClassifier>> classifier =
      TextClassifier::Create(std::move(options));

  EXPECT_EQ(classifier.status().code(), absl::StatusCode::kInvalidArgument);
  EXPECT_THAT(
      classifier.status().message(),
      HasSubstr("ExternalFile must specify at least one of 'file_content', "
                "'file_name', 'file_pointer_meta' or 'file_descriptor_meta'."));
  EXPECT_THAT(classifier.status().GetPayload(kMediaPipeTasksPayload),
              Optional(absl::Cord(absl::StrCat(
                  MediaPipeTasksStatus::kRunnerInitializationError))));
}

TEST_F(TextClassifierTest, CreateFailsWithMissingModel) {
  auto options = std::make_unique<TextClassifierOptions>();
  options->base_options.model_asset_path = GetFullPath(kInvalidModelPath);
  StatusOr<std::unique_ptr<TextClassifier>> classifier =
      TextClassifier::Create(std::move(options));

  EXPECT_EQ(classifier.status().code(), absl::StatusCode::kNotFound);
  EXPECT_THAT(classifier.status().message(),
              HasSubstr("Unable to open file at"));
  EXPECT_THAT(classifier.status().GetPayload(kMediaPipeTasksPayload),
              Optional(absl::Cord(absl::StrCat(
                  MediaPipeTasksStatus::kRunnerInitializationError))));
}

TEST_F(TextClassifierTest, CreateSucceedsWithRegexModel) {
  auto options = std::make_unique<TextClassifierOptions>();
  options->base_options.model_asset_path = GetFullPath(kTestRegexModelPath);
  MP_ASSERT_OK(TextClassifier::Create(std::move(options)));
}

TEST_F(TextClassifierTest, TextClassifierWithBert) {
  auto options = std::make_unique<TextClassifierOptions>();
  options->base_options.model_asset_path = GetFullPath(kTestBertModelPath);
  MP_ASSERT_OK_AND_ASSIGN(std::unique_ptr<TextClassifier> classifier,
                          TextClassifier::Create(std::move(options)));

  TextClassifierResult negative_expected;
  TextClassifierResult positive_expected;

#ifdef _WIN32
  negative_expected.classifications.emplace_back(Classifications{
      /*categories=*/{
          {/*index=*/0, /*score=*/0.956124, /*category_name=*/"negative"},
          {/*index=*/1, /*score=*/0.043875, /*category_name=*/"positive"}},
      /*head_index=*/0,
      /*head_name=*/"probability"});
  positive_expected.classifications.emplace_back(Classifications{
      /*categories=*/{
          {/*index=*/1, /*score=*/0.999951, /*category_name=*/"positive"},
          {/*index=*/0, /*score=*/0.000048, /*category_name=*/"negative"}},
      /*head_index=*/0,
      /*head_name=*/"probability"});
#else
  negative_expected.classifications.emplace_back(Classifications{
      /*categories=*/{{0, 0.963325, "negative"}, {1, 0.036674, "positive"}},
      /*head_index=*/0,
      /*head_name=*/"probability"});
  positive_expected.classifications.emplace_back(Classifications{
      /*categories=*/{{1, 0.9999370, "positive"}, {0, 0.0000629, "negative"}},
      /*head_index=*/0,
      /*head_name=*/"probability"});
#endif  // _WIN32

  MP_ASSERT_OK_AND_ASSIGN(
      TextClassifierResult negative_result,
      classifier->Classify("unflinchingly bleak and desperate"));
  ExpectApproximatelyEqual(negative_result, negative_expected);

  MP_ASSERT_OK_AND_ASSIGN(
      TextClassifierResult positive_result,
      classifier->Classify("it's a charming and often affecting journey"));
  ExpectApproximatelyEqual(positive_result, positive_expected);

  MP_ASSERT_OK(classifier->Close());
}

TEST_F(TextClassifierTest, TextClassifierWithIntInputs) {
  auto options = std::make_unique<TextClassifierOptions>();
  options->base_options.model_asset_path = GetFullPath(kTestRegexModelPath);
  MP_ASSERT_OK_AND_ASSIGN(std::unique_ptr<TextClassifier> classifier,
                          TextClassifier::Create(std::move(options)));
  MP_ASSERT_OK_AND_ASSIGN(TextClassifierResult negative_result,
                          classifier->Classify("What a waste of my time."));
  TextClassifierResult negative_expected;
  negative_expected.classifications.emplace_back(Classifications{
      /*categories=*/{
          {/*index=*/0, /*score=*/0.813130, /*category_name=*/"Negative"},
          {/*index=*/1, /*score=*/0.186870, /*category_name=*/"Positive"}},
      /*head_index=*/0,
      /*head_name=*/"probability"});
  ExpectApproximatelyEqual(negative_result, negative_expected);

  MP_ASSERT_OK_AND_ASSIGN(
      TextClassifierResult positive_result,
      classifier->Classify("This is the best movie I’ve seen in recent years."
                           "Strongly recommend it!"));
  TextClassifierResult positive_expected;
  positive_expected.classifications.emplace_back(Classifications{
      /*categories=*/{
          {/*index=*/1, /*score=*/0.513427, /*category_name=*/"Positive"},
          {/*index=*/0, /*score=*/0.486573, /*category_name=*/"Negative"}},
      /*head_index=*/0,
      /*head_name=*/"probability"});
  ExpectApproximatelyEqual(positive_result, positive_expected);

  MP_ASSERT_OK(classifier->Close());
}

TEST_F(TextClassifierTest, TextClassifierWithStringToBool) {
  auto options = std::make_unique<TextClassifierOptions>();
  options->base_options.model_asset_path = GetFullPath(kStringToBoolModelPath);
  options->base_options.op_resolver = CreateCustomResolver();
  MP_ASSERT_OK_AND_ASSIGN(std::unique_ptr<TextClassifier> classifier,
                          TextClassifier::Create(std::move(options)));
  MP_ASSERT_OK_AND_ASSIGN(TextClassifierResult result,
                          classifier->Classify("hello"));

  // Binary outputs causes flaky ordering, so we compare manually.
  ASSERT_EQ(result.classifications.size(), 1);
  ASSERT_EQ(result.classifications[0].head_index, 0);
  ASSERT_EQ(result.classifications[0].categories.size(), 3);
  ASSERT_EQ(result.classifications[0].categories[0].score, 1);
  ASSERT_LT(result.classifications[0].categories[0].index, 2);  // i.e O or 1.
  ASSERT_EQ(result.classifications[0].categories[1].score, 1);
  ASSERT_LT(result.classifications[0].categories[1].index, 2);  // i.e 0 or 1.
  ASSERT_EQ(result.classifications[0].categories[2].score, 0);
  ASSERT_EQ(result.classifications[0].categories[2].index, 2);
  MP_ASSERT_OK(classifier->Close());
}

TEST_F(TextClassifierTest, BertLongPositive) {
  std::stringstream ss_for_positive_review;
  ss_for_positive_review
      << "it's a charming and often affecting journey and this is a long";
  for (int i = 0; i < kMaxSeqLen; ++i) {
    ss_for_positive_review << " long";
  }
  ss_for_positive_review << " movie review";
  auto options = std::make_unique<TextClassifierOptions>();
  options->base_options.model_asset_path = GetFullPath(kTestBertModelPath);
  MP_ASSERT_OK_AND_ASSIGN(std::unique_ptr<TextClassifier> classifier,
                          TextClassifier::Create(std::move(options)));
  MP_ASSERT_OK_AND_ASSIGN(TextClassifierResult result,
                          classifier->Classify(ss_for_positive_review.str()));
  TextClassifierResult expected;
  std::vector<Category> categories;

// Predicted scores are slightly different on Windows.
#ifdef _WIN32
  categories.push_back(
      {/*index=*/1, /*score=*/0.976686, /*category_name=*/"positive"});
  categories.push_back(
      {/*index=*/0, /*score=*/0.023313, /*category_name=*/"negative"});
#else
  categories.push_back({1, 0.981097, "positive"});
  categories.push_back({0, 0.018902, "negative"});
#endif  // _WIN32

  expected.classifications.emplace_back(
      Classifications{/*categories=*/categories,
                      /*head_index=*/0,
                      /*head_name=*/"probability"});
  ExpectApproximatelyEqual(result, expected);
  MP_ASSERT_OK(classifier->Close());
}

}  // namespace mediapipe::tasks::text::text_classifier

File: mediapipe/tasks/python/test/text/text_classifier_test.py
# Copyright 2022 The MediaPipe Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Tests for text classifier."""

import enum
import os

from absl.testing import absltest
from absl.testing import parameterized

from mediapipe.tasks.python.components.containers import category
from mediapipe.tasks.python.components.containers import classification_result as classification_result_module
from mediapipe.tasks.python.core import base_options as base_options_module
from mediapipe.tasks.python.test import test_utils
from mediapipe.tasks.python.text import text_classifier

TextClassifierResult = classification_result_module.ClassificationResult
_BaseOptions = base_options_module.BaseOptions
_Category = category.Category
_Classifications = classification_result_module.Classifications
_TextClassifier = text_classifier.TextClassifier
_TextClassifierOptions = text_classifier.TextClassifierOptions

_BERT_MODEL_FILE = 'bert_text_classifier.tflite'
_REGEX_MODEL_FILE = 'test_model_text_classifier_with_regex_tokenizer.tflite'
_TEST_DATA_DIR = 'mediapipe/tasks/testdata/text'

_NEGATIVE_TEXT = 'What a waste of my time.'
_POSITIVE_TEXT = ('This is the best movie I’ve seen in recent years.'
                  'Strongly recommend it!')

_BERT_NEGATIVE_RESULTS = TextClassifierResult(
    classifications=[
        _Classifications(
            categories=[
                _Category(
                    index=0,
                    score=0.9995,
                    display_name='',
                    category_name='negative'),
                _Category(
                    index=1,
                    score=0.0005,
                    display_name='',
                    category_name='positive')
            ],
            head_index=0,
            head_name='probability')
    ],
    timestamp_ms=0)
_BERT_POSITIVE_RESULTS = TextClassifierResult(
    classifications=[
        _Classifications(
            categories=[
                _Category(
                    index=1,
                    score=0.9994,
                    display_name='',
                    category_name='positive',
                ),
                _Category(
                    index=0,
                    score=0.0006,
                    display_name='',
                    category_name='negative',
                ),
            ],
            head_index=0,
            head_name='probability',
        )
    ],
    timestamp_ms=0,
)
_REGEX_NEGATIVE_RESULTS = TextClassifierResult(
    classifications=[
        _Classifications(
            categories=[
                _Category(
                    index=0,
                    score=0.81313,
                    display_name='',
                    category_name='Negative'),
                _Category(
                    index=1,
                    score=0.1868704,
                    display_name='',
                    category_name='Positive')
            ],
            head_index=0,
            head_name='probability')
    ],
    timestamp_ms=0)
_REGEX_POSITIVE_RESULTS = TextClassifierResult(
    classifications=[
        _Classifications(
            categories=[
                _Category(
                    index=1,
                    score=0.5134273,
                    display_name='',
                    category_name='Positive'),
                _Category(
                    index=0,
                    score=0.486573,
                    display_name='',
                    category_name='Negative')
            ],
            head_index=0,
            head_name='probability')
    ],
    timestamp_ms=0)


class ModelFileType(enum.Enum):
  FILE_CONTENT = 1
  FILE_NAME = 2


class TextClassifierTest(parameterized.TestCase):

  def setUp(self):
    super().setUp()
    self.model_path = test_utils.get_test_data_path(
        os.path.join(_TEST_DATA_DIR, _BERT_MODEL_FILE))

  def test_create_from_file_succeeds_with_valid_model_path(self):
    # Creates with default option and valid model file successfully.
    with _TextClassifier.create_from_model_path(self.model_path) as classifier:
      self.assertIsInstance(classifier, _TextClassifier)

  def test_create_from_options_succeeds_with_valid_model_path(self):
    # Creates with options containing model file successfully.
    base_options = _BaseOptions(model_asset_path=self.model_path)
    options = _TextClassifierOptions(base_options=base_options)
    with _TextClassifier.create_from_options(options) as classifier:
      self.assertIsInstance(classifier, _TextClassifier)

  def test_create_from_options_fails_with_invalid_model_path(self):
    with self.assertRaisesRegex(
        RuntimeError, 'Unable to open file at /path/to/invalid/model.tflite'):
      base_options = _BaseOptions(
          model_asset_path='/path/to/invalid/model.tflite')
      options = _TextClassifierOptions(base_options=base_options)
      _TextClassifier.create_from_options(options)

  def test_create_from_options_succeeds_with_valid_model_content(self):
    # Creates with options containing model content successfully.
    with open(self.model_path, 'rb') as f:
      base_options = _BaseOptions(model_asset_buffer=f.read())
      options = _TextClassifierOptions(base_options=base_options)
      classifier = _TextClassifier.create_from_options(options)
      self.assertIsInstance(classifier, _TextClassifier)

  @parameterized.parameters(
      (ModelFileType.FILE_NAME, _BERT_MODEL_FILE, _NEGATIVE_TEXT,
       _BERT_NEGATIVE_RESULTS), (ModelFileType.FILE_CONTENT, _BERT_MODEL_FILE,
                                 _NEGATIVE_TEXT, _BERT_NEGATIVE_RESULTS),
      (ModelFileType.FILE_NAME, _BERT_MODEL_FILE, _POSITIVE_TEXT,
       _BERT_POSITIVE_RESULTS), (ModelFileType.FILE_CONTENT, _BERT_MODEL_FILE,
                                 _POSITIVE_TEXT, _BERT_POSITIVE_RESULTS),
      (ModelFileType.FILE_NAME, _REGEX_MODEL_FILE, _NEGATIVE_TEXT,
       _REGEX_NEGATIVE_RESULTS), (ModelFileType.FILE_CONTENT, _REGEX_MODEL_FILE,
                                  _NEGATIVE_TEXT, _REGEX_NEGATIVE_RESULTS),
      (ModelFileType.FILE_NAME, _REGEX_MODEL_FILE, _POSITIVE_TEXT,
       _REGEX_POSITIVE_RESULTS), (ModelFileType.FILE_CONTENT, _REGEX_MODEL_FILE,
                                  _POSITIVE_TEXT, _REGEX_POSITIVE_RESULTS))
  def test_classify(self, model_file_type, model_name, text,
                    expected_classification_result):
    # Creates classifier.
    model_path = test_utils.get_test_data_path(
        os.path.join(_TEST_DATA_DIR, model_name))
    if model_file_type is ModelFileType.FILE_NAME:
      base_options = _BaseOptions(model_asset_path=model_path)
    elif model_file_type is ModelFileType.FILE_CONTENT:
      with open(model_path, 'rb') as f:
        model_content = f.read()
      base_options = _BaseOptions(model_asset_buffer=model_content)
    else:
      # Should never happen
      raise ValueError('model_file_type is invalid.')

    options = _TextClassifierOptions(base_options=base_options)
    classifier = _TextClassifier.create_from_options(options)

    # Performs text classification on the input.
    text_result = classifier.classify(text)
    # Comparing results.
    test_utils.assert_proto_equals(self, text_result.to_pb2(),
                                   expected_classification_result.to_pb2())
    # Closes the classifier explicitly when the classifier is not used in
    # a context.
    classifier.close()

  @parameterized.parameters((ModelFileType.FILE_NAME, _BERT_MODEL_FILE,
                             _NEGATIVE_TEXT, _BERT_NEGATIVE_RESULTS),
                            (ModelFileType.FILE_CONTENT, _BERT_MODEL_FILE,
                             _NEGATIVE_TEXT, _BERT_NEGATIVE_RESULTS))
  def test_classify_in_context(self, model_file_type, model_name, text,
                               expected_classification_result):
    # Creates classifier.
    model_path = test_utils.get_test_data_path(
        os.path.join(_TEST_DATA_DIR, model_name))
    if model_file_type is ModelFileType.FILE_NAME:
      base_options = _BaseOptions(model_asset_path=model_path)
    elif model_file_type is ModelFileType.FILE_CONTENT:
      with open(model_path, 'rb') as f:
        model_content = f.read()
      base_options = _BaseOptions(model_asset_buffer=model_content)
    else:
      # Should never happen
      raise ValueError('model_file_type is invalid.')

    options = _TextClassifierOptions(base_options=base_options)

    with _TextClassifier.create_from_options(options) as classifier:
      # Performs text classification on the input.
      text_result = classifier.classify(text)
      # Comparing results.
      test_utils.assert_proto_equals(self, text_result.to_pb2(),
                                     expected_classification_result.to_pb2())


if __name__ == '__main__':
  absltest.main()
